File: openclaw-studio-frontend/src/services/knowledge.ts
/**
 * 知识库 API 服务
 * 
 * 提供知识库相关的 API 调用接口
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import apiClient from './apiClient';

// 知识库项类型定义
export interface KnowledgeItem {
  path: string;
  title: string;
  content: string;
  category: string;
  tags: string[];
  created_at?: string;
  updated_at?: string;
}

// 搜索响应
export interface SearchResponse {
  results: KnowledgeItem[];
  total: number;
}

/**
 * 搜索知识库
 */
export const useKnowledgeSearchQuery = (
  query: string,
  category?: string,
  tags?: string[]
) => {
  return useQuery<SearchResponse>({
    queryKey: ['knowledge-search', query, category, tags],
    queryFn: async () => {
      const params: Record<string, string> = { q: query };
      if (category) params.category = category;
      if (tags && tags.length > 0) params.tags = tags.join(',');
      
      const response = await apiClient.get('/api/v1/knowledge/search', { params });
      return response.data;
    },
    enabled: !!query && query.length > 0,
  });
};

/**
 * 列出所有模板
 */
export const useTemplatesQuery = () => {
  return useQuery<string[]>({
    queryKey: ['knowledge-templates'],
    queryFn: async () => {
      const response = await apiClient.get('/api/v1/knowledge/templates');
      return response.data;
    },
  });
};

/**
 * 获取模板内容
 */
export const useTemplateQuery = (templateName: string) => {
  return useQuery<{ template_name: string; content: string }>({
    queryKey: ['knowledge-template', templateName],
    queryFn: async () => {
      const response = await apiClient.get(`/api/v1/knowledge/templates/${templateName}`);
      return response.data;
    },
    enabled: !!templateName,
  });
};

/**
 * 列出知识库项
 */
export const useKnowledgeItemsQuery = (category?: string) => {
  return useQuery<KnowledgeItem[]>({
    queryKey: ['knowledge-items', category],
    queryFn: async () => {
      const params = category ? { category } : {};
      const response = await apiClient.get('/api/v1/knowledge/items', { params });
      return response.data;
    },
  });
};

/**
 * 归档案例
 */
export const useArchiveCaseMutation = () => {
  const queryClient = useQueryClient();
  
  return useMutation({
    mutationFn: async (caseId: string) => {
      const response = await apiClient.post(`/api/v1/knowledge/cases/${caseId}/archive`);
      return response.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['knowledge-items'] });
      queryClient.invalidateQueries({ queryKey: ['knowledge-search'] });
    },
  });
};

File: openclaw_studio/api/v1/knowledge.py
"""
知识库 API

提供知识库搜索、模板管理等 API 端点。
"""

from fastapi import APIRouter, HTTPException, Depends, Query
from typing import List, Optional
from pydantic import BaseModel

from openclaw_studio.knowledge_base import KnowledgeBase, KnowledgeItem
from openclaw_core.logger import get_logger

logger = get_logger("openclaw.api.knowledge")

router = APIRouter(prefix="/knowledge", tags=["Knowledge"])


# Pydantic 模型
class KnowledgeItemOut(BaseModel):
    """知识库项输出模型"""
    path: str
    title: str
    content: str
    category: str
    tags: List[str]
    created_at: Optional[str] = None
    updated_at: Optional[str] = None

    model_config = {"from_attributes": True}


class SearchRequest(BaseModel):
    """搜索请求模型"""
    query: str
    category: Optional[str] = None
    tags: Optional[List[str]] = None


class SearchResponse(BaseModel):
    """搜索响应模型"""
    results: List[KnowledgeItemOut]
    total: int


class UpdateItemRequest(BaseModel):
    """更新知识库项请求模型"""
    content: Optional[str] = None
    tags: Optional[List[str]] = None


class TagOperationRequest(BaseModel):
    """标签操作请求模型"""
    tags: List[str]


# 依赖注入
def get_knowledge_base() -> KnowledgeBase:
    """获取知识库实例"""
    return KnowledgeBase()


@router.get("/search", response_model=SearchResponse)
async def search_knowledge(
    q: str = Query(..., description="搜索关键词（支持多关键词，空格分隔）"),
    category: Optional[str] = Query(None, description="类别筛选"),
    tags: Optional[str] = Query(None, description="标签筛选（逗号分隔）"),
    use_regex: bool = Query(False, description="是否使用正则表达式"),
    highlight: bool = Query(False, description="是否高亮匹配内容"),
    knowledge_base: KnowledgeBase = Depends(get_knowledge_base),
):
    """搜索知识库（支持高级搜索）"""
    try:
        tag_list = tags.split(",") if tags else None
        if tag_list:
            tag_list = [tag.strip() for tag in tag_list]
        
        results = knowledge_base.search(
            q, 
            category=category, 
            tags=tag_list,
            use_regex=use_regex,
            highlight=highlight
        )
        
        return SearchResponse(
            results=[KnowledgeItemOut(**item.__dict__) for item in results],
            total=len(results),
        )
    except Exception as e:
        logger.error(f"搜索知识库失败: {e}")
        raise HTTPException(status_code=500, detail=f"Search failed: {str(e)}")


@router.get("/templates", response_model=List[str])
async def list_templates(
    knowledge_base: KnowledgeBase = Depends(get_knowledge_base),
):
    """列出所有模板"""
    try:
        return knowledge_base.list_templates()
    except Exception as e:
        logger.error(f"获取模板列表失败: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to list templates: {str(e)}")


@router.get("/templates/{template_name}")
async def get_template(
    template_name: str,
    knowledge_base: KnowledgeBase = Depends(get_knowledge_base),
):
    """获取模板内容"""
    try:
        content = knowledge_base.get_template(template_name)
        if not content:
            raise HTTPException(status_code=404, detail="Template not found")
        return {"template_name": template_name, "content": content}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取模板失败: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get template: {str(e)}")


@router.get("/items", response_model=List[KnowledgeItemOut])
async def list_items(
    category: Optional[str] = Query(None, description="类别筛选"),
    knowledge_base: KnowledgeBase = Depends(get_knowledge_base),
):
    """列出知识库项"""
    try:
        items = knowledge_base.list_items(category=category)
        return [KnowledgeItemOut(**item.__dict__) for item in items]
    except Exception as e:
        logger.error(f"列出知识库项失败: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to list items: {str(e)}")


@router.put("/items/{item_path:path}")
async def update_item(
    item_path: str,
    request: UpdateItemRequest,
    knowledge_base: KnowledgeBase = Depends(get_knowledge_base),
):
    """更新知识库项（编辑内容或标签）"""
    try:
        success = knowledge_base.update_item(
            item_path=item_path,
            content=request.content,
            tags=request.tags
        )
        if not success:
            raise HTTPException(status_code=404, detail="Item not found or update failed")
        return {"message": "Item updated successfully", "path": item_path}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"更新知识库项失败: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to update item: {str(e)}")


@router.post("/items/{item_path:path}/tags")
async def add_tags(
    item_path: str,
    request: TagOperationRequest,
    knowledge_base: KnowledgeBase = Depends(get_knowledge_base),
):
    """添加标签到知识库项"""
    try:
        success = knowledge_base.add_tags(item_path, request.tags)
        if not success:
            raise HTTPException(status_code=404, detail="Item not found")
        return {"message": "Tags added successfully", "path": item_path}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"添加标签失败: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to add tags: {str(e)}")


@router.delete("/items/{item_path:path}/tags")
async def remove_tags(
    item_path: str,
    request: TagOperationRequest,
    knowledge_base: KnowledgeBase = Depends(get_knowledge_base),
):
    """从知识库项中移除标签"""
    try:
        success = knowledge_base.remove_tags(item_path, request.tags)
        if not success:
            raise HTTPException(status_code=404, detail="Item not found")
        return {"message": "Tags removed successfully", "path": item_path}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"移除标签失败: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to remove tags: {str(e)}")


@router.post("/cases/{case_id}/archive")
async def archive_case(
    case_id: str,
    knowledge_base: KnowledgeBase = Depends(get_knowledge_base),
):
    """归档案例到知识库"""
    try:
        from pathlib import Path
        from openclaw_studio.case_manager import CaseManager
        
        case_manager = CaseManager()
        case = case_manager.get_case(case_id)
        if not case:
            raise HTTPException(status_code=404, detail="Case not found")
        
        # 获取案例目录
        case_dir = Path("cases") / case_id
        if not case_dir.exists():
            raise HTTPException(status_code=404, detail="Case directory not found")
        
        success = knowledge_base.archive_case(case_id, case_dir)
        if not success:
            raise HTTPException(status_code=500, detail="Failed to archive case")
        
        return {"message": f"Case {case_id} archived successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"归档案例失败: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to archive case: {str(e)}")
